compiler/src/iree/compiler/Codegen/Common/ForallToFor.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "llvm/ADT/SmallVectorExtras.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Pass/Pass.h"
+
+#include <cassert>
+
+#define DEBUG_TYPE "iree-codegen-forall-to-for"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_FORALLTOFORPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+static LogicalResult forallToForLoop(RewriterBase &rewriter,
+                                     scf::ForallOp forallOp) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(forallOp);
+
+  // Check pre-condition, the buildBody function below needs to know how to
+  // convert all the ops inside `scf.forall.in_parallel`.
+  auto terminator = forallOp.getTerminator();
+  if (!all_of(terminator.getYieldingOps(),
+              llvm::IsaPred<tensor::ParallelInsertSliceOp>)) {
+    return failure();
+  }
+
+  // Build `scf.for` loop nest.
+  SmallVector<Value> lbs = forallOp.getLowerBound(rewriter);
+  SmallVector<Value> ubs = forallOp.getUpperBound(rewriter);
+  SmallVector<Value> steps = forallOp.getStep(rewriter);
+  SmallVector<Value> iterArgs = forallOp.getOutputs();
+  auto buildBody = [&](OpBuilder &builder, Location loc, ValueRange ivs,
+                       ValueRange args) -> scf::ValueVector {
+    // Inline `scf.forall` body excluding terminating
+    // `scf.forall.in_parallel` op.
+    IRMapping map;
+    map.map(forallOp.getInductionVars(), ivs);
+    map.map(forallOp.getRegionOutArgs(), args);
+    for (auto &op : forallOp.getBody()->without_terminator()) {
+      builder.clone(op, map);
+    }
+
+    // Convert + inline the contents of `scf.forall.in_parallel` terminator.
+    SmallVector<Value> yieldedValues;
+    for (auto &yieldOp : terminator.getYieldingOps()) {
+      assert(isa<tensor::ParallelInsertSliceOp>(yieldOp) &&
+             "unsupported operation in in_parallel region");
```

Comment:
optional nit: I'd drop the assertion because it is in lambda and we already have the check above.